### ISSUE TITLE
Harness: 3 native-CUDA smoke tests fail on ambiguous model.io.token_input (fixture omits IO declaration)

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnx-genai#1284]
## Summary

Three CUDA-gated native-decode tests fail **only under `ONNX_GENAI_RUN_CUDA_SMOKE=1`** with `cannot resolve model.io.token_input ... 3 ports match`. This is a **test-harness** problem, not a product bug: the synthetic decoder fixture presents three structurally identical input ports without declaring `model.io.token_input`, and the product's IO-role resolver correctly refuses to guess among them.

Filing per request to stop treating these as permanently "known pre-existing" — three separate agents have now excused them in one day. **Not asking for a fix here**, only a durable record of which side (harness) owns them.

## Affected tests (`crates/onnx-genai-engine/src/native_decode/tests.rs`)

- `native_cuda_capture_replay_is_bit_exact_and_refreshes_decode_inputs`
- `native_cuda_symbolic_batch_aux_captures_bit_exact`
- `native_cuda_symbolic_total_seq_aux_declines_capture_but_decodes_eagerly`

## Exact error (RTX 4060, CUDA 13 runtime)

```
Error: cannot resolve model.io.token_input from tensor shape because 3 ports match:
[("input_ids", Int64, [1, 1]), ("attention_mask", Int64, [1, -1]), ("position_ids", Int64, [1, 1])];
declare the exact graph port in model.io.token_input
```

## Root cause — HARNESS

The three tests build the synthetic graph through `capture_safe_cuda_decoder` → `build_cuda_decoder` → `build_cuda_decoder_with_fixed_state(.., fixed_state = false)`. In the `fixed_state = false` branch the session is constructed with **`NativeDecodeSession::from_session_with_cuda_options(...)`, which passes no `model.io` metadata**. The graph exposes three rank‑2 `Int64` inputs — `input_ids [1,1]`, `attention_mask [1,-1]`, `position_ids [1,1]` — all of which satisfy the `is_rank_one_or_two_sequence` structural predicate for the token input.

`onnx_genai_ort::io_roles::resolve_port` (in `crates/onnx-genai-ort/src/io_roles.rs`) is **deliberately** conservative: its module contract is *"Metadata is authoritative. Unambiguous tensor-shape signals are the only permitted fallback; graph port names are never interpreted."* With three structural matches and no `model.io.token_input` declaration, it returns the actionable error above. **This is correct, intended product behavior.**

### Why this is harness, not product

- Real models resolve fine under the same resolver: the smoke test `native_cuda_vmm_kv_grows_in_place_and_commits_more_granules` loads `qwen2.5-0.5b-int4` via `load_with_resolved_io` and passes — real decoder exports either declare `model.io.token_input` or carry distinguishing shapes.
- The harness's **`fixed_state = true`** path passes `tiny_decoder_io()` (which declares the ports explicitly) and constructs without error; #1264's and #1266's metadata-less tests use that path and pass.
- Only the `fixed_state = false` builder omits the IO declaration, which is what makes the fixture — not the product — ambiguous.

## Suggested fix (for whoever picks this up; out of scope here)

Declare the token input on the `fixed_state = false` builder path too — e.g. thread an `io` with `token_input = "input_ids"` (as the `fixed_state = true` path already does via `tiny_decoder_io()`), or give the fixture distinguishing shapes so structural resolution is unambiguous. No product change to `resolve_port` is warranted.

## Reproduce

```powershell
$nv = "$env:LOCALAPPDATA\anaconda3\Lib\site-packages\nvidia"
$env:PATH = "$nv\cu13\bin\x86_64;$nv\cudnn\bin;$env:PATH"
$env:ONNX_GENAI_RUN_CUDA_SMOKE = "1"
cargo test -p onnx-genai-engine --features "cuda,native-backend" --lib `
  native_cuda_capture_replay_is_bit_exact_and_refreshes_decode_inputs `
  -- --test-threads=1 --nocapture
```

Hardware: RTX 4060 (8 GB, driver 591.55, CUDA 13.1 runtime, WDDM). These three are the only smoke failures; the rest of the engine lib suite is **556 passed / 3 failed / 2 ignored** with smoke on, **559 / 0 / 2** with smoke off.

**Comment by justinchuby** [comment]
Diagnosis is **incomplete** — measured on RTX 4060 (CUDA 13.1, driver 591.55), synthetic decoder fixture, `ONNX_GENAI_RUN_CUDA_SMOKE=1`, `--test-threads=1`.

**First-order diagnosis confirmed (falsified A/B).** Declaring `model.io` via `tiny_decoder_io()` in `build_cuda_decoder` (mirroring the `fixed_state=true` path) resolves the `model.io.token_input` ambiguity. Reverting it reproduces the exact error on all three tests: `cannot resolve model.io.token_input ... 3 ports match [input_ids, attention_mask, position_ids]`. Fix in draft PR #1458 (product `resolve_port` untouched).

**But that only fixes one of the three.** `native_cuda_symbolic_total_seq_...` → PASS. The two capture-enabled tests (`native_cuda_capture_replay_...`, `native_cuda_symbolic_batch_...`) still fail, now for a **second, independent reason** — identical measured decline at construction:
```
predicate `persistent_inputs_have_fixed_logical_shapes` declined capture:
past_key_values.0.key (physical [1,1,16,1] vs logical [1,1,0,1]),
past_key_values.0.value (physical [1,1,16,1] vs logical [1,1,0,1])
expose a growing logical prefix instead of fixed capacity
```

**Root cause of the second blocker:** the fixture routes growable KV through `Cast` (`present.0.key = Cast(past_key)`), not a capacity-form attention kernel. `binding_consumers_use_physical_capacity` recognizes only `GroupQueryAttention`, mask-driven `Attention`, and `pkg.nxrt::IndexShare` as physical-capacity KV consumers — `Cast` is not one — so the KV binding exposes a growing logical prefix (`logical[2]=0 ≠ physical[2]=max_len`) and CUDA-graph capture is correctly declined. There is no KV auto-detection (`load.rs` bails without `kv_inputs`), so the full `tiny_decoder_io()` is mandatory and the decline is unavoidable with this fixture.

**Recommendation:** fully greening the two capture tests needs the fixture's KV to feed a capacity-form attention op (real GQA/Attention that executes on CUDA), which is a substantial fixture rewrite that changes what these structural tests exercise. I did **not** do this, and did **not** weaken the `captures=1`/bit-exact assertions to force green (same hard-stop discipline as the ULP sweep). Leaving #1284 open; PR #1458 lands only the confirmed-correct `token_input` fix as the necessary first step.

Full lib smoke suite with the fix: **565 passed / 2 failed / 2 ignored** (569 total).

**Comment by justinchuby** [comment]
**Update — partial fix landed (#1458, squash-merged), two tests gated, fixture rewrite remains.**

Landed:
1. `build_cuda_decoder` now declares `model.io` via `tiny_decoder_io()` (mirroring the working `fixed_state=true` path), resolving the `token_input` ambiguity. Product `resolve_port` untouched. This **fully fixes `native_cuda_symbolic_total_seq`**.
2. `native_cuda_capture_replay` and `native_cuda_symbolic_batch` are gated with a substantive `#[ignore]` (assertions untouched). They assert `captures=1`/`graph_enabled`, which this fixture structurally cannot deliver: growable KV is routed through `Cast`, not a capacity-form attention op, so `binding_consumers_use_physical_capacity` correctly declines CUDA-graph capture (`persistent_inputs_have_fixed_logical_shapes`). Two permanently-red tests train reviewers to skim past red output — the exact cost we paid all day — so they are ignored with the cause and the unskip condition recorded in-source.

Verification (RTX 4060, CUDA 13.1, driver 591.55, `ONNX_GENAI_RUN_CUDA_SMOKE=1`, `--test-threads=1`):
- Before: **565 passed / 2 failed / 2 ignored**
- After: **565 passed / 0 failed / 4 ignored**

**Remaining work (keeping this issue open):** unskipping the two capture tests requires rewriting the fixture so its KV feeds a **real CUDA attention kernel** (`GroupQueryAttention`, mask-driven `Attention`, or `pkg.nxrt::IndexShare`) that binds the cache at physical capacity — the only consumers `binding_consumers_use_physical_capacity` accepts. That attention op must actually execute on CUDA during the smoke run, and it changes what the tests exercise (structural F1/F2 capture analysis over a realistic decode graph). This is a genuine piece of work, not a five-minute fix — do **not** weaken the `captures=1`/bit-exact assertions to force green.
